pkg/manifest/jsonmanifest/jsonmanifest_test.go

```diff
@@ -0,0 +1,243 @@
+
+	"github.com/ethersphere/bee/pkg/manifest"
+	"github.com/ethersphere/bee/pkg/manifest/jsonmanifest"
+	"github.com/ethersphere/bee/pkg/swarm"
+	"github.com/ethersphere/bee/pkg/swarm/test"
+)
+
+var testCases = []testCase{
+	{
+		name:    "empty-manifest",
+		entries: nil,
+	},
+	{
+		name: "one-entry",
+		entries: []e{
+			{
+				reference: test.RandomAddress(),
+				name:      "entry-1",
+				headers:   http.Header{},
+				path:      "",
+			},
+		},
+	},
+	{
+		name: "two-entries",
+		entries: []e{
+			{
+				reference: test.RandomAddress(),
+				name:      "entry-1.txt",
+				headers:   http.Header{"Content-Type": {"text/plain; charset=utf-8"}},
+				path:      "",
+			},
+			{
+				reference: test.RandomAddress(),
+				name:      "entry-2.png",
+				headers:   http.Header{"Content-Type": {"image/png"}},
+				path:      "",
+			},
+		},
+	},
+	{
+		name: "nested-entries",
+		entries: []e{
+			{
+				reference: test.RandomAddress(),
+				name:      "robots.txt",
+				headers:   http.Header{"Content-Type": {"text/plain; charset=utf-8"}},
+				path:      "text",
+			},
+			{
+				reference: test.RandomAddress(),
+				name:      "1.png",
+				headers:   http.Header{"Content-Type": {"image/png"}},
+				path:      "img",
+			},
+			{
+				reference: test.RandomAddress(),
+				name:      "2.jpg",
+				headers:   http.Header{"Content-Type": {"image/jpg"}},
+				path:      "img",
+			},
+			{
+				reference: test.RandomAddress(),
+				name:      "readme.md",
+				headers:   http.Header{"Content-Type": {"text/markdown; charset=UTF-8"}},
+				path:      "",
+			},
+		},
+	},
+}
+
+// TestEntries verifies that manifests behave as expected when adding and removing entries.
+// It also tests the Length and Entry functions.
+func TestEntries(t *testing.T) {
+	// get non-trivial test case
+	tc := testCases[len(testCases)-1] // last test case
+	if len(tc.entries) < 2 {
+		t.Fatal("cannot test manifest entries using a test case with less than 2 entries")
+	}
+
+	m := jsonmanifest.NewManifest()
+	if m.Length() != 0 {
+		t.Fatalf("expected length to be %d, but is %d instead", 0, m.Length())
+	}
+
+	// add and check all entries
+	for i, e := range tc.entries {
+		entry := jsonmanifest.NewEntry(
+			e.reference,
+			e.name,
+			e.headers,
+		)
+		path := filepath.Join(e.path, e.name)
+		m.Add(path, entry)
+
+		// verify length change
+		if m.Length() != i+1 {
+			t.Fatalf("expected length to be %d, but is %d instead", i+1, m.Length())
+		}
+		// check retrieved entry
+		verifyEntry(t, m, entry, path)
+	}
+	// save current manifest length
+	manifestLen := m.Length()
+
+	// create new entry to replace existing one
+	lastEntry := tc.entries[len(tc.entries)-1]
+	entry := jsonmanifest.NewEntry(
+		test.RandomAddress(),
+		lastEntry.name,
+		lastEntry.headers,
+	)
+
+	// replace manifest entry by adding to the same path
+	path := filepath.Join(lastEntry.path, lastEntry.name)
+	m.Add(path, entry)
+
+	// length should not have changed
+	if m.Length() != manifestLen {
```

Comment:
these repetitions of `m.Length() ....` could be a helper function:
```go
func checkManifestLength(t *testing.T, m *manifest, count int){
if m.length != count { t.Fatal(yada yada) } ...
}

Reply:
extracted `checkLength` helper func. thanks